File: src/main/java/com/dv/telegram/command/GetStatistics.java
package com.dv.telegram.command;

import com.dv.telegram.BotStatistics;
import com.dv.telegram.WikiBot;
import org.apache.commons.lang3.StringUtils;

import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

public class GetStatistics extends BasicBotCommand {

    private final DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy - HH:mm:ss");

    @Override
    public String getName() {
        return GetStatistics.class.getSimpleName();
    }

    @Override
    public String getDescription(String botName) {
        return String.format(
            "`%s %s` — получить статистику по работе бота с момента текущего запуска инстанса.",
            botName,
            getCommandText()
        );
    }

    @Override
    public String getDefaultCommandName() {
        return "/getStats";
    }

    @Override
    public String getResponse(String text, WikiBot bot) {
        BotStatistics statistics = bot.getStatistics();

        List<String> statisticsLines = List.of(
            getStatisticsLine("Время старта бота", statistics.startTime),
            getStatisticsLine("Успешных запросов", statistics.getSuccessfulRequestsCountWithPercentage()),
            getStatisticsLine("Неуспешных запросов", statistics.getFailedRequestsCountWithPercentage()),
            getStatisticsLine("Всего запросов", statistics.getTotalCallsWithPercentage()),
            getStatisticsLine("Вызовов специальных команд", statistics.specialCommandsCount),
            getStatisticsLine("Всего запросов (вместе со специальными командами)", statistics.getTotalCallsWithSpecialCommands())
        );

        return StringUtils.join(statisticsLines, "\n");
    }

    private String getStatisticsLine(String name, long count) {
        return String.format("— %s: %d", name, count);
    }

    private String getStatisticsLine(String name, String value) {
        return String.format("— %s: %s", name, value);
    }

    private String getStatisticsLine(String name, ZonedDateTime time) {
        return String.format("— %s: %s", name, dateTimeFormatter.format(time));
    }
}

File: src/main/kotlin/com/dv/telegram/command/GetStatistics.kt
package com.dv.telegram.command

import com.dv.telegram.WikiBot
import org.telegram.telegrambots.meta.api.objects.Update
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

class GetStatistics : BasicBotCommand() {
    override val name: String = javaClass.simpleName

    override fun getDescription(bot: WikiBot) =
        "`${bot.botName} $commandText` — получить статистику по работе бота с момента текущего запуска инстанса."

    override val defaultCommandName = "/getStats"

    override fun getResponse(text: String, bot: WikiBot, update: Update): String {
        val statistics = bot.statistics

        val statisticsLines = listOf(
            getStatisticsLine("Время старта бота", statistics.startTime),
            getStatisticsLine("Успешных запросов", statistics.successfulRequestsCountWithPercentage),
            getStatisticsLine("Неуспешных запросов", statistics.failedRequestsCountWithPercentage),
            getStatisticsLine("Всего запросов", statistics.totalCallsWithPercentage),
            getStatisticsLine("Вызовов специальных команд", statistics.specialCommandsCount),
            getStatisticsLine("Всего запросов (вместе со специальными командами)", statistics.totalCallsWithSpecialCommands)
        )

        return statisticsLines.joinToString("\n")
    }

    companion object {
        private val dateTimeFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy - HH:mm:ss")

        fun getStatisticsLine(name: String, count: Long) =
            "— $name: $count"

        fun getStatisticsLine(name: String, value: String) =
            "— $name: $value"

        fun getStatisticsLine(name: String, time: ZonedDateTime) =
            "— $name: ${dateTimeFormatter.format(time)}"
    }
}
